### DISCUSSION TITLE
Best way to make script exit on (most) errors?

### DISCUSSION THREAD

**Discussion by tiymat** [lmorg/murex#756] [Q&A] [answered]
Hey everyone, wanted to run something by you. My situation is that I'm working on a script. I generally want it to fail (stop executing and exit with non-zero exit status) whenever a command terminates with a non-zero exit status. But there are a couple of commands in the script I want to be able to exit with any status without stopping the script.

My current approach is to make the script `runmode trypipe module` and encase most of the script in `try` blocks, with the exception of the commands I want to be able to exit non-zero.

e.g.

```
try {
    do_thing_that_shouldnt_error
}
do_thing_that_can_error
try {
    do_thing_that_shouldnt_error
}
```
This seems to work but is kind of clumsy (95% of the script is encase in try blocks) and I'm wondering if I'm missing a better approach. Any ideas?

**Comment by lmorg** [comment]
It is clumsy. For a while now, I've been considering adding something like `unsafe {}`. It would be a very easy addition because I've already written prototype code called `evil`. But I've always been able to work around that limitation by wrapping things in `private` functions.

Using this new block it would look something like this:
```
do_thing_that_shouldnt_error
unsafe { do_thing_that_can_error }
do_thing_that_shouldnt_error
```

The idea is that `unsafe` / `evil` / whatever we call it, that function would gobble up the exit code and always return `0`, regardless of the actual error returned from any of the code that runs within the "unsafe" block.

Would this be a worthwhile solution for yourself? Most of the code is already sat there ready to be turned on, I just haven't had a strong enough justification for enabling that feature before now.

**Reply by tiymat** [reply]
I like that idea a lot. If I understand correctly, to get my desired behavior I'd still need to enclose the whole script in a `try` block, but the commands in the `unsafe` block wouldn't trigger the termination of the `try` block?

I'm wondering if there's a step further we could take to avoid needing the outer `try` block, possibly as an optional 3rd `runmode` parameter, that says basically "treat this module/function as if it were enclosed in a `try` block"?

**Reply by lmorg** [reply]
> I like that idea a lot. If I understand correctly, to get my desired behavior I'd still need to enclose the whole script in a `try` block, but the commands in the `unsafe` block wouldn't trigger the termination of the `try` block?

Yes, that's exactly the behaviour of the `unsafe` block. Though you wouldn't need to enclose the whole script in a `try` block _per se_ because using `runmode try function` (or `try module`) has the same effect as a `try {}` block. eg

```
function footer {
    # This is an example

    runmode try function

    do_thing_that_shouldnt_error
    unsafe { do_thing_that_can_error }
    do_thing_that_shouldnt_error
}
```
is equivalent to:
```
function footer {
    # This is an example

    try {

        do_thing_that_shouldnt_error
        unsafe { do_thing_that_can_error }
        do_thing_that_shouldnt_error

    }
}
```

> I'm wondering if there's a step further we could take to avoid needing the outer `try` block, possibly as an optional 3rd `runmode` parameter, that says basically "treat this module/function as if it were enclosed in a `try` block"?

I'm not sure what you mean here but `runmode` does support both `try` and `trypipe` modes and at function and module scope: https://murex.rocks/commands/runmode.html

If that doesn't cover what you mean, then are you able to provide an example?

**Reply by tiymat** [reply]
I'm likely misunderstanding something then. I was testing this by running a script starting with `runmode trypipe module` and observed that the behavior I wanted (that the script stop on an error) wasn't happening unless it was wrapped in a `try` block. But when I put it in a function using `runmode trypipe function` it does seem to stop. Is this intended behavior?

You can see below, test-mod.mx outputs the 'finished!' string after the command fails, while test-func does not. The behavior doesn't change if I change the scripts to use the `try` runmode rather than `trypipe`.

![runmode_test](https://github.com/lmorg/murex/assets/138939221/4e9d0d95-6a75-4438-9554-01d8b014753a)

**Reply by tiymat** [reply]
I forgot to mention, when I use `runmode trypipe module` and enclose the code in a `try` block it also works as expected and exits immediately on error, which is where I was coming from with my question about adding an arg to `runmode`.
![runmode_test1](https://github.com/lmorg/murex/assets/138939221/a8498c25-b42a-44d9-a80b-fe3bb5c60e8a)

**Reply by lmorg** [reply, marked as answer]
ahhh, now i see the issue. `module` is intended for modules in that it descends through to functions. eg

```
#!/usr/bin/murex

runmode trypipe module

function foobar {
    false

    out hello world
}

foobar
```
In this, the function `foobar` will exit on `false` because `foobar` inherits `trypipe` from the module.

What should _also_ happen is any code within the root scope of that module should also inherit `trypipe` -- but as you've seen from your own code, that isn't what is happening. 

Technically that's by design... but in practice that design sucks. So I'll look to get that changed in the next release.

**Reply by lmorg** [reply]
oh I clean forgot to add: if you add `-trypipe` flag to `murex`, it runs everything in that session as `trypipe` by default.

eg
```
#!/usr/bin/murex -trypipe

not_a_real_command
out "finished!"
```
will error as you expect.

I should still fix `runmode ... module` to do the same too -- since that is the behaviour people would expect.

**Reply by tiymat** [reply]
Makes sense, thank you!

**Reply by lmorg** [reply]
`unsafe` and the `runmode trypipe module` fix are both in `develop` branch now, will be released as part of v5.2

https://dev.murex.rocks/commands/unsafe.html
